File: android/app/src/main/java/com/bwqr/mavinote/models/TraitHelpers.kt
package com.bwqr.mavinote.models

import com.novi.bincode.BincodeDeserializer
import com.novi.serde.Deserializer

class TraitHelpers {
    companion object {
        fun<T> deserializeOption(bytes: ByteArray, deserialize: (deserializer: Deserializer) -> T): T? {
            val deserializer = BincodeDeserializer(bytes)
            val tag = deserializer.deserialize_option_tag()

            return if (!tag) {
                null
            } else {
                deserialize(deserializer)
            }
        }

        fun<T> deserializeList(bytes: ByteArray, deserialize: (deserializer: Deserializer) -> T): List<T> {
            val deserializer = BincodeDeserializer(bytes)
            val items = mutableListOf<T>()

            for (index in 0 until deserializer.deserialize_len()) {
                items.add(deserialize(deserializer))
            }

            return items
        }
    }
}

File: ios/MaviNote/Models/TraitHelpers.swift
import Foundation
import Serde

func deserializeList<T>(_ deserializer: Deserializer, deserialize: (_ deserializer: Deserializer) throws -> T) throws -> [T] {
    var items: [T] = []
    let len = try deserializer.deserialize_len()

    for _ in 0...len {
        items.append(try deserialize(deserializer))
    }

    return items
}
